keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+
+### Goals
+- improve readiness times for clusters with a large number of encrypted resources
+- reduce the likelihood of hitting the KMS rate limit
+- enable fully automated key rotation for the latest key
+- improve KMS plugin health check reliability
+- improve observability of envelop operations between kube-apiserver, KMS plugins and KMS
+- if this v2 API reaches GA in release N, the existing v1beta1 gRPC API will be deprecated at release N and removed at release N+3 (the existing key rotation dance of using multiple providers will be used to migrate from v1beta1 to v2)
+
+### Non-Goals
+- Prevent KMS rate limiting
+- Recovery when KMS KEK is deleted
+- Using the proposed transaction ID for audit logging
+
+## Proposal
+
+Performance, Health Check, Observability and Rotation:
+- Support key hierarchy in KMS plugin that generates local KEK
+- Expand `EncryptionConfiguration` to support a new KMSv2 configuration
+- Add v2alpha1 `KeyManagementService` proto service contract in Kubernetes to include
+    - `currentKeyID` and metadata to support key rotation
+    - `currentKeyID`: the KMS Key ID, stable identifier, changed to trigger key rotation and storage migration
+    - metadata: structured data, can contain the encrypted local KEK, can be used for debugging, recovery, opaque to API server, stored unencrypted, etc. Validation similar to how K8s labels are validated today. Labels have good size limits and restrictions today.
+    - A status request and response periodically (order of minutes) returns `version`, `healthz`, and `currentKeyID`
+    - The `currentKeyID` in status can be used on decrypt operations to compare and validate the key ID stored in the DEK cache and the latest `EncryptResponse` `currentKeyID` to detect if an object is stale in terms of storage migration
+    - Generate a new UID for each envelope operation in kube-apiserver
+    - Add a new UID field to `EncryptRequest` and `DecryptRequest`
+- Add support for hot reload of the `EncryptionConfiguration`:
+    - Watch on the `EncryptionConfiguration`
+    - When changes are detected, process the `EncryptionConfiguration` resource, and add new transformers and update existing ones atomically.
+    - If there is an issue with creating or updating any of the transformers, retain the current configuration in the kube-apiserver and generate an error in logs.
```

Comment:
I am skeptical. Is this an instance of "adding more and more layers to a system intended to avert disaster often makes catastrophe all the more likely to happen"?

Reply:
Not sure I follow - all this is really trying to do is make it so that if you do need change your encryption config, you do not need to restart the API server and lose the DEK cache.